# 00. Table of Content
### 01. Importing Libraries & Datasets
### 02. Describe Datasets
#### D2 - 2015
#### D2 - 2016
#### D2 - 2017
#### D2 - 2018
#### D2 - 2019
#### D2 - 2020
#### D2 - 2021
#### D2 - 2022
#### D2 - 2023
#### D1 - Full 2005 - 2022
### 03. Append D2 Datasets
### 04. Export D2 Datasets

## 01. Importing Libraries & Datasets

In [42]:
import pandas as pd
import numpy as np
import os

In [43]:
# Create path
path = r'C:\Users\krist\21.02.2024 World Happiness Report Analysis'

In [44]:
# Import data set 1
d2_2015 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2015 report (2012-2014).csv'), index_col = False)

In [45]:
# Import data set 2
d2_2016 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2016 report (2013-2015).csv'), index_col = False)

In [46]:
# Import data set 3
d2_2017 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2017 report (2014-2016).csv'), index_col = False)

In [47]:
# Import data set 4
d2_2018 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2018 report (2015-2017).csv'), index_col = False)

In [48]:
# Import data set 5
d2_2019 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2019 report (2016-2018).csv'), index_col = False)

In [49]:
# Import data set 6
d2_2020 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2020 report (2017-2019).csv'), index_col = False)

In [50]:
# Import data set 7
d2_2021 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2021 report (2018-2020).csv'), index_col = False)

In [51]:
# Import data set 8
d2_2022 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2022 report (2019-2021).csv'), index_col = False)

In [52]:
# Import data set 9
d2_2023 = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd2_2023 report (2020-2022).csv'), index_col = False)

In [53]:
# Import data set 10
d1_full = pd.read_csv(os.path.join(path, '02 Data', 'Sourced Data', 'd1_full_score_data (2005-2022).csv'), index_col = False)

## 02. Describe Datasets

### D2 - 2015

In [54]:
# View data set
d2_2015

year  rank      country            regional indicator  ladder score  \
0    2015     1  Switzerland                Western Europe         7.587   
1    2015     2      Iceland                Western Europe         7.561   
2    2015     3      Denmark                Western Europe         7.527   
3    2015     4       Norway                Western Europe         7.522   
4    2015     5       Canada         North America and ANZ         7.427   
..    ...   ...          ...                           ...           ...   
153  2015   154       Rwanda            Sub-Saharan Africa         3.465   
154  2015   155        Benin            Sub-Saharan Africa         3.340   
155  2015   156        Syria  Middle East and North Africa         3.006   
156  2015   157      Burundi            Sub-Saharan Africa         2.905   
157  2015   158         Togo            Sub-Saharan Africa         2.839   

     log GDP per capita  social support  healthy life expectancy  \
0              1.396505        1.349505                 0.941432   
1              1.302324        1.402231                 0.947844   
2              1.325478        1.360581                 0.874641   
3              1.458997        1.330955                 0.885209   
4              1.326292        1.322608                 0.905631   
..                  ...             ...                      ...   
153            0.222078        0.773701                 0.428643   
154            0.286655        0.353862                 0.319098   
155            0.663199        0.474890                 0.721932   
156            0.015303        0.415867                 0.223961   
157            0.208676        0.139945                 0.284430   

     freedom to make life choices  generosity  perceptions of corruption  \
0                        0.665573    0.296775                   0.419777   
1                        0.628772    0.436297                   0.141451   
2                        0.649384    0.341386                   0.483573   
3                        0.669732    0.346989                   0.365034   
4                        0.632968    0.458109                   0.329573   
..                            ...         ...                        ...   
153                      0.592007    0.226279                   0.551911   
154                      0.484496    0.182600                   0.080100   
155                      0.156841    0.471790                   0.189060   
156                      0.118496    0.197269                   0.100620   
157                      0.364528    0.166809                   0.107313   

     dystopia + residual  
0               2.517383  
1               2.702013  
2               2.492042  
3               2.465312  
4               2.451756  
..                   ...  
153             0.670418  
154             1.633279  
155             0.328579  
156             1.833019  
157             1.567258  

[158 rows x 12 columns]

In [55]:
# Describe data set
d2_2015.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   158.0  158.000000    158.000000          158.000000      158.000000   
mean   2015.0   79.500000      5.375734            0.846137        0.991046   
std       0.0   45.754781      1.145010            0.403121        0.272369   
min    2015.0    1.000000      2.839000            0.000000        0.000000   
25%    2015.0   40.250000      4.526000            0.545806        0.856820   
50%    2015.0   79.500000      5.232500            0.910242        1.029509   
75%    2015.0  118.750000      6.243750            1.158448        1.214408   
max    2015.0  158.000000      7.587000            1.690415        1.402231   

       healthy life expectancy  freedom to make life choices  generosity  \
count               158.000000                    158.000000  158.000000   
mean                  0.630260                      0.428615    0.237295   
std                   0.247078                      0.150693    0.126685   
min                   0.000000                      0.000000    0.000000   
25%                   0.439183                      0.328329    0.150552   
50%                   0.696709                      0.435517    0.216128   
75%                   0.811015                      0.549091    0.309882   
max                   1.025247                      0.669732    0.795885   

       perceptions of corruption  dystopia + residual  
count                 158.000000           158.000000  
mean                    0.143421             2.098976  
std                     0.120034             0.553550  
min                     0.000000             0.328579  
25%                     0.061674             1.759406  
50%                     0.107223             2.095414  
75%                     0.180256             2.462415  
max                     0.551911             3.602143

In [56]:
d2_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   year                          158 non-null    int64  
 1   rank                          158 non-null    int64  
 2   country                       158 non-null    object 
 3   regional indicator            158 non-null    object 
 4   ladder score                  158 non-null    float64
 5   log GDP per capita            158 non-null    float64
 6   social support                158 non-null    float64
 7   healthy life expectancy       158 non-null    float64
 8   freedom to make life choices  158 non-null    float64
 9   generosity                    158 non-null    float64
 10  perceptions of corruption     158 non-null    float64
 11  dystopia + residual           158 non-null    float64
dtypes: float64(8), int64(2), object(2)
memory usage: 14.9+ KB


Summary 2015 Report data
Poll data used (2012-2014)
158 Coutries included
Minimum Happiness Score 2.83 Togo
Maximum Happiness Score 7.58 Switzerland
Avg. Happiness Score 

### D2 - 2016

In [57]:
# View data set top 5 entries
d2_2016.head()

year  rank      country regional indicator  ladder score  \
0  2016     1      Denmark     Western Europe         7.526   
1  2016     2  Switzerland     Western Europe         7.509   
2  2016     3      Iceland     Western Europe         7.501   
3  2016     4       Norway     Western Europe         7.498   
4  2016     5      Finland     Western Europe         7.413   

   log GDP per capita  social support  healthy life expectancy  \
0            1.441784        1.163745                 0.795040   
1            1.527334        1.145236                 0.863029   
2            1.426655        1.183259                 0.867333   
3            1.577440        1.126896                 0.795789   
4            1.405979        1.134640                 0.810914   

   freedom to make life choices  generosity  perceptions of corruption  \
0                      0.579413    0.361710                   0.444532   
1                      0.585572    0.280825                   0.412033   
2                      0.566242    0.476784                   0.149747   
3                      0.596086    0.378948                   0.357758   
4                      0.571037    0.254919                   0.410040   

   dystopia + residual  whisker-high  whisker-low  
0             2.739390         7.592        7.460  
1             2.694632         7.590        7.428  
2             2.831373         7.669        7.333  
3             2.664645         7.575        7.421  
4             2.825965         7.475        7.351

In [58]:
# Describe data set
d2_2016.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   157.0  157.000000    157.000000          157.000000      157.000000   
mean   2016.0   79.000000      5.382185            0.953880        0.793621   
std       0.0   45.466105      1.141674            0.412595        0.266706   
min    2016.0    1.000000      2.905000            0.000000        0.000000   
25%    2016.0   40.000000      4.404000            0.670235        0.641837   
50%    2016.0   79.000000      5.314000            1.027797        0.841420   
75%    2016.0  118.000000      6.269000            1.279637        1.021524   
max    2016.0  157.000000      7.526000            1.824266        1.183259   

       healthy life expectancy  freedom to make life choices  generosity  \
count               157.000000                    157.000000  157.000000   
mean                  0.557619                      0.370993    0.242635   
std                   0.229349                      0.145507    0.133756   
min                   0.000000                      0.000000    0.000000   
25%                   0.382906                      0.257483    0.154574   
50%                   0.596592                      0.397465    0.222450   
75%                   0.729932                      0.484529    0.311846   
max                   0.952773                      0.608483    0.819711   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 157.000000           157.000000    157.000000   
mean                    0.137624             2.325807      5.481975   
std                     0.111038             0.542220      1.136493   
min                     0.000000             0.817891      3.078000   
25%                     0.061263             2.031708      4.465000   
50%                     0.105465             2.290740      5.419000   
75%                     0.175544             2.664645      6.434000   
max                     0.505209             3.837715      7.669000   

       whisker-low  
count   157.000000  
mean      5.282395  
std       1.148043  
min       2.732000  
25%       4.327000  
50%       5.237000  
75%       6.154000  
max       7.460000

In [59]:
d2_2016.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2017

In [60]:
# View data set top 5 entries
d2_2017.head()

year  rank      country regional indicator  ladder score  \
0  2017     1       Norway     Western Europe         7.537   
1  2017     2      Denmark     Western Europe         7.522   
2  2017     3      Iceland     Western Europe         7.504   
3  2017     4  Switzerland     Western Europe         7.494   
4  2017     5      Finland     Western Europe         7.469   

   log GDP per capita  social support  healthy life expectancy  \
0               1.616           1.534                    0.797   
1               1.482           1.551                    0.793   
2               1.481           1.611                    0.834   
3               1.565           1.517                    0.858   
4               1.444           1.540                    0.809   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.635       0.362                      0.316   
1                         0.626       0.355                      0.401   
2                         0.627       0.476                      0.154   
3                         0.620       0.291                      0.367   
4                         0.618       0.245                      0.383   

   dystopia + residual  whisker-high  whisker-low  
0                2.277         7.594        7.480  
1                2.314         7.582        7.462  
2                2.323         7.622        7.386  
3                2.277         7.562        7.426  
4                2.430         7.528        7.410

In [61]:
# Describe data set
d2_2017.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   155.0  155.000000    155.000000          155.000000      155.000000   
mean   2017.0   78.000000      5.354019            0.984716        1.188871   
std       0.0   44.888751      1.131230            0.420809        0.287245   
min    2017.0    1.000000      2.693000            0.000000        0.000000   
25%    2017.0   39.500000      4.505500            0.663500        1.042500   
50%    2017.0   78.000000      5.279000            1.065000        1.254000   
75%    2017.0  116.500000      6.101500            1.318000        1.414000   
max    2017.0  155.000000      7.537000            1.871000        1.611000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               155.000000                    155.000000  155.000000   
mean                  0.551342                      0.408761    0.246884   
std                   0.237035                      0.150001    0.134814   
min                   0.000000                      0.000000    0.000000   
25%                   0.370000                      0.304000    0.154000   
50%                   0.606000                      0.437000    0.232000   
75%                   0.723000                      0.516500    0.323500   
max                   0.949000                      0.658000    0.838000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 155.000000           155.000000    155.000000   
mean                    0.123097             1.850206      5.452342   
std                     0.101629             0.500061      1.118527   
min                     0.000000             0.378000      2.865000   
25%                     0.057000             1.591500      4.608500   
50%                     0.090000             1.833000      5.370000   
75%                     0.153500             2.145000      6.194500   
max                     0.464000             3.117000      7.622000   

       whisker-low  
count   155.000000  
mean      5.255697  
std       1.145044  
min       2.521000  
25%       4.375000  
50%       5.193000  
75%       6.006500  
max       7.480000

In [62]:
d2_2017.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2018

In [63]:
# View data set top 5 entries
d2_2018.head()

year  rank      country regional indicator  ladder score  \
0  2018     1      Finland     Western Europe         7.632   
1  2018     2       Norway     Western Europe         7.594   
2  2018     3      Denmark     Western Europe         7.555   
3  2018     4      Iceland     Western Europe         7.495   
4  2018     5  Switzerland     Western Europe         7.487   

   log GDP per capita  social support  healthy life expectancy  \
0               1.305           1.592                    0.874   
1               1.456           1.582                    0.861   
2               1.351           1.590                    0.868   
3               1.343           1.644                    0.914   
4               1.420           1.549                    0.927   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.681       0.192                      0.393   
1                         0.686       0.286                      0.340   
2                         0.683       0.284                      0.408   
3                         0.677       0.353                      0.138   
4                         0.660       0.256                      0.357   

   dystopia + residual  whisker-high  whisker-low  
0                2.595         7.695        7.569  
1                2.383         7.657        7.530  
2                2.370         7.623        7.487  
3                2.426         7.593        7.398  
4                2.320         7.570        7.405

In [64]:
# Describe data set
d2_2018.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   156.0  156.000000    156.000000          156.000000      156.000000   
mean   2018.0   78.500000      5.375878            0.887417        1.216571   
std       0.0   45.177428      1.119507            0.382573        0.300380   
min    2018.0    1.000000      2.905000            0.000000        0.000000   
25%    2018.0   39.750000      4.453750            0.616250        1.076750   
50%    2018.0   78.500000      5.378000            0.949500        1.261500   
75%    2018.0  117.250000      6.168500            1.197750        1.463000   
max    2018.0  156.000000      7.632000            1.649000        1.644000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.598026                      0.456981    0.181571   
std                   0.247925                      0.162750    0.098805   
min                   0.000000                      0.000000    0.000000   
25%                   0.422250                      0.358250    0.109500   
50%                   0.644000                      0.494000    0.174000   
75%                   0.777250                      0.580000    0.242250   
max                   1.030000                      0.724000    0.598000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 156.000000           156.000000    156.000000   
mean                    0.112474             1.922891      5.478654   
std                     0.096362             0.513760      1.107522   
min                     0.000000             0.292000      3.074000   
25%                     0.051000             1.654250      4.589750   
50%                     0.082000             1.908500      5.477500   
75%                     0.139000             2.270250      6.260000   
max                     0.457000             2.961000      7.695000   

       whisker-low  
count   156.000000  
mean      5.273090  
std       1.132544  
min       2.735000  
25%       4.344750  
50%       5.284500  
75%       6.051250  
max       7.569000

In [65]:
d2_2018.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2019

In [66]:
# View data set top 5 entries
d2_2019.head()

year  rank      country regional indicator  ladder score  \
0  2019     1      Finland     Western Europe         7.769   
1  2019     2      Denmark     Western Europe         7.600   
2  2019     3       Norway     Western Europe         7.554   
3  2019     4      Iceland     Western Europe         7.494   
4  2019     5  Netherlands     Western Europe         7.488   

   log GDP per capita  social support  healthy life expectancy  \
0               1.340           1.587                    0.986   
1               1.383           1.573                    0.996   
2               1.488           1.582                    1.028   
3               1.380           1.624                    1.026   
4               1.396           1.522                    0.999   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.596       0.153                      0.393   
1                         0.592       0.252                      0.410   
2                         0.603       0.271                      0.341   
3                         0.591       0.354                      0.118   
4                         0.557       0.322                      0.298   

   dystopia + residual  whisker-high  whisker-low  
0                2.714         7.830        7.708  
1                2.393         7.667        7.534  
2                2.241         7.616        7.492  
3                2.401         7.613        7.374  
4                2.393         7.542        7.433

In [67]:
# Describe data set
d2_2019.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   156.0  156.000000    156.000000          156.000000      156.000000   
mean   2019.0   78.500000      5.407096            0.905147        1.208814   
std       0.0   45.177428      1.113120            0.398389        0.299191   
min    2019.0    1.000000      2.853000            0.000000        0.000000   
25%    2019.0   39.750000      4.544500            0.602750        1.055750   
50%    2019.0   78.500000      5.379500            0.960000        1.271500   
75%    2019.0  117.250000      6.184500            1.232500        1.452500   
max    2019.0  156.000000      7.769000            1.684000        1.624000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.725244                      0.392571    0.184846   
std                   0.242124                      0.143289    0.095254   
min                   0.000000                      0.000000    0.000000   
25%                   0.547750                      0.308000    0.108750   
50%                   0.789000                      0.417000    0.177500   
75%                   0.881750                      0.507250    0.248250   
max                   1.141000                      0.631000    0.566000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 156.000000           156.000000    156.000000   
mean                    0.110603             1.879936      5.511968   
std                     0.094538             0.528797      1.098572   
min                     0.000000             0.183000      3.005000   
25%                     0.047000             1.513250      4.665000   
50%                     0.085500             1.917000      5.471500   
75%                     0.141250             2.267500      6.276000   
max                     0.453000             3.049000      7.830000   

       whisker-low  
count   156.000000  
mean      5.302122  
std       1.128566  
min       2.701000  
25%       4.436500  
50%       5.285500  
75%       6.070500  
max       7.708000

In [68]:
d2_2019.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2020

In [69]:
# View data set top 5 entries
d2_2020.head()

year  rank      country regional indicator  ladder score  \
0  2020     1      Finland     Western Europe        7.8087   
1  2020     2      Denmark     Western Europe        7.6456   
2  2020     3  Switzerland     Western Europe        7.5599   
3  2020     4      Iceland     Western Europe        7.5045   
4  2020     5       Norway     Western Europe        7.4880   

   log GDP per capita  social support  healthy life expectancy  \
0            1.285190        1.499526                 0.961271   
1            1.326949        1.503449                 0.979333   
2            1.390774        1.472403                 1.040533   
3            1.326502        1.547567                 1.000843   
4            1.424207        1.495173                 1.008072   

   freedom to make life choices  generosity  perceptions of corruption  \
0                      0.662317    0.159670                   0.477857   
1                      0.665040    0.242793                   0.495260   
2                      0.628954    0.269056                   0.407946   
3                      0.661981    0.362330                   0.144541   
4                      0.670201    0.287985                   0.434101   

   dystopia + residual  whisker-high  whisker-low  
0             2.762835      7.869766     7.747634  
1             2.432741      7.711245     7.579955  
2             2.350267      7.628528     7.491272  
3             2.460688      7.621347     7.387653  
4             2.168266      7.556281     7.419719

In [70]:
# Describe data set
d2_2020.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   153.0  153.000000     153.00000          153.000000      153.000000   
mean   2020.0   77.000000       5.47324            0.868771        1.155607   
std       0.0   44.311398       1.11227            0.372416        0.286866   
min    2020.0    1.000000       2.56690            0.000000        0.000000   
25%    2020.0   39.000000       4.72410            0.575862        0.986718   
50%    2020.0   77.000000       5.51500            0.918549        1.203987   
75%    2020.0  115.000000       6.22850            1.169229        1.387139   
max    2020.0  153.000000       7.80870            1.536676        1.547567   

       healthy life expectancy  freedom to make life choices  generosity  \
count               153.000000                    153.000000  153.000000   
mean                  0.692869                      0.463583    0.189375   
std                   0.254094                      0.141172    0.100401   
min                   0.000000                      0.000000    0.000000   
25%                   0.495443                      0.381457    0.115006   
50%                   0.759818                      0.483293    0.176745   
75%                   0.867249                      0.576665    0.255510   
max                   1.137814                      0.693270    0.569814   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 153.000000           153.000000    153.000000   
mean                    0.130718             1.972317      5.578175   
std                     0.113097             0.563638      1.096823   
min                     0.000000             0.257241      2.628270   
25%                     0.055805             1.629928      4.826248   
50%                     0.098435             2.046272      5.607728   
75%                     0.163064             2.350267      6.363886   
max                     0.533162             3.440810      7.869766   

       whisker-low  
count   153.000000  
mean      5.368304  
std       1.128631  
min       2.505530  
25%       4.603149  
50%       5.430644  
75%       6.138881  
max       7.747634

In [71]:
d2_2020.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2021

In [72]:
# View data set top 5 entries
d2_2021.head()

year  rank      country regional indicator  ladder score  \
0  2021     1      Finland     Western Europe         7.842   
1  2021     2      Denmark     Western Europe         7.620   
2  2021     3  Switzerland     Western Europe         7.571   
3  2021     4      Iceland     Western Europe         7.554   
4  2021     5  Netherlands     Western Europe         7.464   

   log GDP per capita  social support  healthy life expectancy  \
0               1.446           1.106                    0.741   
1               1.502           1.108                    0.763   
2               1.566           1.079                    0.816   
3               1.482           1.172                    0.772   
4               1.501           1.079                    0.753   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.691       0.124                      0.481   
1                         0.686       0.208                      0.485   
2                         0.653       0.204                      0.413   
3                         0.698       0.293                      0.170   
4                         0.647       0.302                      0.384   

   dystopia + residual  whisker-high  whisker-low  
0                3.253         7.904        7.780  
1                2.868         7.687        7.552  
2                2.839         7.643        7.500  
3                2.967         7.670        7.438  
4                2.798         7.518        7.410

In [73]:
# Describe data set
d2_2021.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   149.0  149.000000    149.000000          149.000000      149.000000   
mean   2021.0   75.000000      5.532839            0.977161        0.793315   
std       0.0   43.156691      1.073924            0.404740        0.258871   
min    2021.0    1.000000      2.523000            0.000000        0.000000   
25%    2021.0   38.000000      4.852000            0.666000        0.647000   
50%    2021.0   75.000000      5.534000            1.025000        0.832000   
75%    2021.0  112.000000      6.255000            1.323000        0.996000   
max    2021.0  149.000000      7.842000            1.751000        1.172000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               149.000000                    149.000000  149.000000   
mean                  0.520161                      0.498711    0.178047   
std                   0.213019                      0.137888    0.098270   
min                   0.000000                      0.000000    0.000000   
25%                   0.357000                      0.409000    0.105000   
50%                   0.571000                      0.514000    0.164000   
75%                   0.665000                      0.603000    0.239000   
max                   0.897000                      0.716000    0.541000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 149.000000           149.000000    149.000000   
mean                    0.135141             2.430329      5.648007   
std                     0.114361             0.537645      1.054330   
min                     0.000000             0.648000      2.596000   
25%                     0.060000             2.138000      4.991000   
50%                     0.101000             2.509000      5.625000   
75%                     0.174000             2.794000      6.344000   
max                     0.547000             3.482000      7.904000   

       whisker-low  
count   149.000000  
mean      5.417631  
std       1.094879  
min       2.449000  
25%       4.706000  
50%       5.413000  
75%       6.128000  
max       7.780000

In [74]:
d2_2021.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2022

In [75]:
# View data set top 5 entries
d2_2022.head()

year  rank      country regional indicator  ladder score  \
0  2022     1      Finland     Western Europe         7.821   
1  2022     2      Denmark     Western Europe         7.636   
2  2022     3      Iceland     Western Europe         7.557   
3  2022     4  Switzerland     Western Europe         7.512   
4  2022     5  Netherlands     Western Europe         7.415   

   log GDP per capita  social support  healthy life expectancy  \
0               1.892           1.258                    0.775   
1               1.953           1.243                    0.777   
2               1.936           1.320                    0.803   
3               2.026           1.226                    0.822   
4               1.945           1.206                    0.787   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.736       0.109                      0.534   
1                         0.719       0.188                      0.532   
2                         0.718       0.270                      0.191   
3                         0.677       0.147                      0.461   
4                         0.651       0.271                      0.419   

   dystopia + residual  whisker-high  whisker-low  
0                2.518         7.886        7.756  
1                2.226         7.710        7.563  
2                2.320         7.651        7.464  
3                2.153         7.586        7.437  
4                2.137         7.471        7.359

In [76]:
# Describe data set
d2_2022.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   146.0  146.000000    146.000000          146.000000      146.000000   
mean   2022.0   73.500000      5.553575            1.410445        0.905863   
std       0.0   42.290661      1.086843            0.421663        0.280122   
min    2022.0    1.000000      2.404000            0.000000        0.000000   
25%    2022.0   37.250000      4.888750            1.095500        0.732000   
50%    2022.0   73.500000      5.568500            1.445500        0.957500   
75%    2022.0  109.750000      6.305000            1.784750        1.114250   
max    2022.0  146.000000      7.821000            2.209000        1.320000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               146.000000                    146.000000  146.000000   
mean                  0.586171                      0.517226    0.147377   
std                   0.176336                      0.145859    0.082799   
min                   0.000000                      0.000000    0.000000   
25%                   0.463250                      0.440500    0.089000   
50%                   0.621500                      0.543500    0.132500   
75%                   0.719750                      0.626000    0.197750   
max                   0.942000                      0.740000    0.468000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 146.000000           146.000000    146.000000   
mean                    0.154781             1.831808      5.673589   
std                     0.127514             0.534994      1.065621   
min                     0.000000             0.187000      2.469000   
25%                     0.068250             1.555250      5.006250   
50%                     0.119500             1.894500      5.680000   
75%                     0.198500             2.153000      6.448750   
max                     0.587000             2.844000      7.886000   

       whisker-low  
count   146.000000  
mean      5.433568  
std       1.109380  
min       2.339000  
25%       4.754750  
50%       5.453000  
75%       6.190000  
max       7.756000

In [77]:
d2_2022.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D2 - 2023

In [78]:
# View data set top 5 entries
d2_2023.head()

year  rank      country            regional indicator  ladder score  \
0  2023     1      Finland                Western Europe         7.804   
1  2023     2      Denmark                Western Europe         7.586   
2  2023     3      Iceland                Western Europe         7.530   
3  2023     4       Israel  Middle East and North Africa         7.473   
4  2023     5  Netherlands                Western Europe         7.403   

   log GDP per capita  social support  healthy life expectancy  \
0               1.888           1.585                    0.535   
1               1.949           1.548                    0.537   
2               1.926           1.620                    0.559   
3               1.833           1.521                    0.577   
4               1.942           1.488                    0.545   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.772       0.126                      0.535   
1                         0.734       0.208                      0.525   
2                         0.738       0.250                      0.187   
3                         0.569       0.124                      0.158   
4                         0.672       0.251                      0.394   

   dystopia + residual  whisker-high  whisker-low  
0                2.363         7.875        7.733  
1                2.084         7.667        7.506  
2                2.250         7.625        7.434  
3                2.691         7.535        7.411  
4                2.110         7.460        7.346

In [79]:
# Describe data set
d2_2023.describe()

year        rank  ladder score  log GDP per capita  social support  \
count   137.0  137.000000    137.000000          137.000000      137.000000   
mean   2023.0   69.000000      5.539796            1.406985        1.156212   
std       0.0   39.692569      1.139929            0.432963        0.326322   
min    2023.0    1.000000      1.859000            0.000000        0.000000   
25%    2023.0   35.000000      4.724000            1.099000        0.962000   
50%    2023.0   69.000000      5.684000            1.449000        1.227000   
75%    2023.0  103.000000      6.334000            1.798000        1.401000   
max    2023.0  137.000000      7.804000            2.200000        1.620000   

       healthy life expectancy  freedom to make life choices  generosity  \
count               136.000000                    137.000000  137.000000   
mean                  0.366176                      0.540000    0.148474   
std                   0.156691                      0.149501    0.076053   
min                   0.000000                      0.000000    0.000000   
25%                   0.248500                      0.455000    0.097000   
50%                   0.389500                      0.557000    0.137000   
75%                   0.487500                      0.656000    0.199000   
max                   0.702000                      0.772000    0.422000   

       perceptions of corruption  dystopia + residual  whisker-high  \
count                 137.000000           136.000000    137.000000   
mean                    0.145898             1.777838      5.666526   
std                     0.126723             0.504390      1.117421   
min                     0.000000            -0.110000      1.923000   
25%                     0.060000             1.555250      4.980000   
50%                     0.111000             1.848500      5.797000   
75%                     0.187000             2.078750      6.441000   
max                     0.561000             2.955000      7.875000   

       whisker-low  
count   137.000000  
mean      5.412971  
std       1.163724  
min       1.795000  
25%       4.496000  
50%       5.529000  
75%       6.243000  
max       7.733000

In [80]:
d2_2023.dtypes

year                              int64
rank                              int64
country                          object
regional indicator               object
ladder score                    float64
log GDP per capita              float64
social support                  float64
healthy life expectancy         float64
freedom to make life choices    float64
generosity                      float64
perceptions of corruption       float64
dystopia + residual             float64
whisker-high                    float64
whisker-low                     float64
dtype: object

### D1 - Full 2005 - 2022

In [81]:
# View data set top 15 entries
d1_full.head(15)

country name  year  life Ladder  log GDP per capita  social support  \
0   Afghanistan  2008        3.724               7.350           0.451   
1   Afghanistan  2009        4.402               7.509           0.552   
2   Afghanistan  2010        4.758               7.614           0.539   
3   Afghanistan  2011        3.832               7.581           0.521   
4   Afghanistan  2012        3.783               7.661           0.521   
5   Afghanistan  2013        3.572               7.680           0.484   
6   Afghanistan  2014        3.131               7.671           0.526   
7   Afghanistan  2015        3.983               7.654           0.529   
8   Afghanistan  2016        4.220               7.650           0.559   
9   Afghanistan  2017        2.662               7.648           0.491   
10  Afghanistan  2018        2.694               7.631           0.508   
11  Afghanistan  2019        2.375               7.640           0.420   
12  Afghanistan  2021        2.436               7.324           0.454   
13  Afghanistan  2022        1.281                 NaN           0.228   
14      Albania  2007        4.634               9.122           0.821   

    healthy life expectancy at birth  freedom to make life choices  \
0                             50.500                         0.718   
1                             50.800                         0.679   
2                             51.100                         0.600   
3                             51.400                         0.496   
4                             51.700                         0.531   
5                             52.000                         0.578   
6                             52.300                         0.509   
7                             52.600                         0.389   
8                             52.925                         0.523   
9                             53.250                         0.427   
10                            53.575                         0.374   
11                            53.900                         0.394   
12                            54.550                         0.394   
13                            54.875                         0.368   
14                            66.760                         0.529   

    generosity  perceptions of corruption  positive affect  negative affect  
0        0.168                      0.882            0.414            0.258  
1        0.191                      0.850            0.481            0.237  
2        0.121                      0.707            0.517            0.275  
3        0.164                      0.731            0.480            0.267  
4        0.238                      0.776            0.614            0.268  
5        0.063                      0.823            0.547            0.273  
6        0.106                      0.871            0.492            0.375  
7        0.082                      0.881            0.491            0.339  
8        0.044                      0.793            0.501            0.348  
9       -0.119                      0.954            0.435            0.371  
10      -0.091                      0.928            0.385            0.405  
11      -0.106                      0.924            0.324            0.502  
12      -0.081                      0.946            0.179            0.607  
13         NaN                      0.733            0.206            0.576  
14      -0.010                      0.875            0.489            0.246

In [82]:
# Describe data set
d1_full.describe()

year  life Ladder  log GDP per capita  social support  \
count  2199.000000  2199.000000         2179.000000     2186.000000   
mean   2014.161437     5.479227            9.389760        0.810681   
std       4.718736     1.125527            1.153402        0.120953   
min    2005.000000     1.281000            5.527000        0.228000   
25%    2010.000000     4.647000            8.500000        0.747000   
50%    2014.000000     5.432000            9.499000        0.836000   
75%    2018.000000     6.309500           10.373500        0.905000   
max    2022.000000     8.019000           11.664000        0.987000   

       healthy life expectancy at birth  freedom to make life choices  \
count                       2145.000000                   2166.000000   
mean                          63.294582                      0.747847   
std                            6.901104                      0.140137   
min                            6.720000                      0.258000   
25%                           59.120000                      0.656250   
50%                           65.050000                      0.770000   
75%                           68.500000                      0.859000   
max                           74.475000                      0.985000   

        generosity  perceptions of corruption  positive affect  \
count  2126.000000                2083.000000      2175.000000   
mean      0.000091                   0.745208         0.652148   
std       0.161079                   0.185835         0.105913   
min      -0.338000                   0.035000         0.179000   
25%      -0.112000                   0.688000         0.572000   
50%      -0.023000                   0.800000         0.663000   
75%       0.092000                   0.869000         0.738000   
max       0.703000                   0.983000         0.884000   

       negative affect  
count      2183.000000  
mean          0.271493  
std           0.086872  
min           0.083000  
25%           0.208000  
50%           0.261000  
75%           0.323000  
max           0.705000

In [83]:
d1_full.dtypes

country name                         object
year                                  int64
life Ladder                         float64
log GDP per capita                  float64
social support                      float64
healthy life expectancy at birth    float64
freedom to make life choices        float64
generosity                          float64
perceptions of corruption           float64
positive affect                     float64
negative affect                     float64
dtype: object

### Check for Missing values

In [84]:
# Group dataset by year and count countries
d1_countrycount = d1_full.groupby('year')['country name'].count()

In [85]:
# countries participating per year
d1_countrycount

year
2005     27
2006     89
2007    102
2008    110
2009    114
2010    124
2011    146
2012    141
2013    136
2014    144
2015    142
2016    141
2017    147
2018    141
2019    143
2020    116
2021    122
2022    114
Name: country name, dtype: int64

In [86]:
d1_countrycount.to_clipboard()

In [87]:
# Group dataset by country and count years
d1_yearcount = d1_full.groupby('country name').agg({'year':['count']})

In [88]:
d1_yearcount_sorted = d1_yearcount.sort_values(('year', 'count'),ascending=False)

In [89]:
# Years countries participated in
d1_yearcount_sorted

year
             count
country name      
Lebanon         17
Indonesia       17
Kyrgyzstan      17
Kenya           17
Kazakhstan      17
...            ...
Maldives         1
Guyana           1
Oman             1
Suriname         1
Cuba             1

[165 rows x 1 columns]

In [90]:
d1_yearcount_sorted.to_clipboard()

In [91]:
# Limit D1 to countries that have observations in 2017 - 2020
# Create Crosstab for countries and years
crosstab = pd.crosstab(d1_full['country name'], d1_full['year'], dropna=False)

In [92]:
crosstab.to_clipboard()

In [93]:
# Filter out the years 2017-2022
filtered_years = crosstab.loc[:, 2017:2022]

In [94]:
# Find countries with 0 entries for those years
countries_with_zero_entries = filtered_years[filtered_years.prod(axis=1) == 0].index.tolist()

In [95]:
countries_with_zero_entries

['Afghanistan',
 'Algeria',
 'Angola',
 'Armenia',
 'Azerbaijan',
 'Bahrain',
 'Belarus',
 'Belize',
 'Bhutan',
 'Bosnia and Herzegovina',
 'Botswana',
 'Burkina Faso',
 'Burundi',
 'Central African Republic',
 'Chad',
 'China',
 'Comoros',
 'Congo (Kinshasa)',
 'Cuba',
 'Czechia',
 'Djibouti',
 'Eswatini',
 'Ethiopia',
 'Gambia',
 'Guatemala',
 'Guyana',
 'Haiti',
 'Honduras',
 'Hong Kong S.A.R. of China',
 'Iceland',
 'Iraq',
 'Jamaica',
 'Kuwait',
 'Laos',
 'Lesotho',
 'Liberia',
 'Libya',
 'Luxembourg',
 'Madagascar',
 'Malawi',
 'Malaysia',
 'Maldives',
 'Mauritania',
 'Montenegro',
 'Mozambique',
 'Myanmar',
 'Niger',
 'Nigeria',
 'Oman',
 'Pakistan',
 'Panama',
 'Paraguay',
 'Philippines',
 'Qatar',
 'Rwanda',
 'Serbia',
 'Sierra Leone',
 'Singapore',
 'Slovakia',
 'Somalia',
 'Somaliland region',
 'South Africa',
 'South Sudan',
 'Sri Lanka',
 'State of Palestine',
 'Sudan',
 'Suriname',
 'Syria',
 'Tajikistan',
 'Togo',
 'Trinidad and Tobago',
 'Turkey',
 'Turkmenistan',
 'Uga

In [96]:
count_zero_entries = sum(1 for country in countries_with_zero_entries)

In [97]:
count_zero_entries

76

In [98]:
# Exclude those countries from the original dataframe d1_full
d1_filtered = d1_full[~d1_full['country name'].isin(countries_with_zero_entries)]

In [99]:
# drop entries for the years before 2017 from filtered list
d1_filtered = d1_filtered[d1_filtered['year'] >= 2017]

In [100]:
d1_filtered.describe()

year  life Ladder  log GDP per capita  social support  \
count   534.000000   534.000000          526.000000      533.000000   
mean   2019.500000     5.835041            9.746715        0.829268   
std       1.709426     1.039937            1.007235        0.115950   
min    2017.000000     2.179000            5.527000        0.366000   
25%    2018.000000     5.098250            9.121750        0.770000   
50%    2019.500000     5.956000            9.875000        0.868000   
75%    2021.000000     6.516000           10.635750        0.920000   
max    2022.000000     7.889000           11.625000        0.974000   

       healthy life expectancy at birth  freedom to make life choices  \
count                        522.000000                    532.000000   
mean                          66.038218                      0.807776   
std                            5.203859                      0.103772   
min                           52.150000                      0.423000   
25%                           63.825000                      0.740750   
50%                           66.425000                      0.821000   
75%                           70.650000                      0.887000   
max                           74.475000                      0.985000   

       generosity  perceptions of corruption  positive affect  negative affect  
count  525.000000                 511.000000       530.000000       530.000000  
mean    -0.008181                   0.716730         0.661911         0.279472  
std      0.157709                   0.193423         0.098388         0.085364  
min     -0.338000                   0.151000         0.263000         0.083000  
25%     -0.125000                   0.645000         0.598000         0.212000  
50%     -0.032000                   0.780000         0.674500         0.269000  
75%      0.107000                   0.853000         0.738750         0.333750  
max      0.552000                   0.961000         0.826000         0.569000

#### Crosstab check shows 89 eligible countries to compare within the time frame of 2017 - 2022. 76 countries will be excluded from further analysis. As will years before 2017.

In [101]:
# Append 2023 and 2020 d2 datasets and create crosstab to identify missing countries
d2_comp = pd.concat([d2_2023, d2_2020])

In [102]:
d2_comp

year  rank                   country            regional indicator  \
0    2023     1                   Finland                Western Europe   
1    2023     2                   Denmark                Western Europe   
2    2023     3                   Iceland                Western Europe   
3    2023     4                    Israel  Middle East and North Africa   
4    2023     5               Netherlands                Western Europe   
..    ...   ...                       ...                           ...   
148  2020   149  Central African Republic            Sub-Saharan Africa   
149  2020   150                    Rwanda            Sub-Saharan Africa   
150  2020   151                  Zimbabwe            Sub-Saharan Africa   
151  2020   152               South Sudan            Sub-Saharan Africa   
152  2020   153               Afghanistan                    South Asia   

     ladder score  log GDP per capita  social support  \
0          7.8040            1.888000        1.585000   
1          7.5860            1.949000        1.548000   
2          7.5300            1.926000        1.620000   
3          7.4730            1.833000        1.521000   
4          7.4030            1.942000        1.488000   
..            ...                 ...             ...   
148        3.4759            0.041072        0.000000   
149        3.3123            0.343243        0.522876   
150        3.2992            0.425564        1.047835   
151        2.8166            0.289083        0.553279   
152        2.5669            0.300706        0.356434   

     healthy life expectancy  freedom to make life choices  generosity  \
0                   0.535000                      0.772000    0.126000   
1                   0.537000                      0.734000    0.208000   
2                   0.559000                      0.738000    0.250000   
3                   0.577000                      0.569000    0.124000   
4                   0.545000                      0.672000    0.251000   
..                       ...                           ...         ...   
148                 0.000000                      0.292814    0.253513   
149                 0.572383                      0.604088    0.235705   
150                 0.375038                      0.377405    0.151349   
151                 0.208809                      0.065609    0.209935   
152                 0.266052                      0.000000    0.135235   

     perceptions of corruption  dystopia + residual  whisker-high  whisker-low  
0                     0.535000             2.363000      7.875000     7.733000  
1                     0.525000             2.084000      7.667000     7.506000  
2                     0.187000             2.250000      7.625000     7.434000  
3                     0.158000             2.691000      7.535000     7.411000  
4                     0.394000             2.110000      7.460000     7.346000  
..                         ...                  ...           ...          ...  
148                   0.028265             2.860198      3.701658     3.250141  
149                   0.485542             0.548445      3.415053     3.209547  
150                   0.080929             0.841031      3.414202     3.184198  
151                   0.111157             1.378751      3.027516     2.605684  
152                   0.001226             1.507236      2.628270     2.505530  

[290 rows x 14 columns]

In [103]:
# Create Crosstab
crosstab_d2 = pd.crosstab(d2_comp['country'], d2_comp['year'], dropna=False)

In [104]:
crosstab_d2

year         2020  2023
country                
Afghanistan     1     1
Albania         1     1
Algeria         1     1
Argentina       1     1
Armenia         1     1
...           ...   ...
Venezuela       1     1
Vietnam         1     1
Yemen           1     0
Zambia          1     1
Zimbabwe        1     1

[156 rows x 2 columns]

In [105]:
# Find countries with 0 entries for those years
countries_with_zero_entries_d2 = crosstab_d2[crosstab_d2.prod(axis=1) == 0].index.tolist()

In [106]:
countries_with_zero_entries_d2

['Azerbaijan',
 'Belarus',
 'Burundi',
 'Central African Republic',
 'Czech Republic',
 'Czechia',
 'Haiti',
 'Kuwait',
 'Lesotho',
 'Libya',
 'Macedonia',
 'Maldives',
 'North Cyprus',
 'North Macedonia',
 'Palestinian Territories',
 'Rwanda',
 'South Sudan',
 'State of Palestine',
 'Swaziland',
 'Trinidad and Tobago',
 'Turkmenistan',
 'Yemen']

In [107]:
# Exclude those countries from the d2_comp dataframe
d2_comp_filtered = d2_comp[~d2_comp['country'].isin(countries_with_zero_entries_d2)]

In [108]:
d2_comp_filtered['year'].value_counts()

year
2023    134
2020    134
Name: count, dtype: int64

In [109]:
d2_comp_filtered.shape

(268, 14)

In [110]:
# Create filtered single version for d2_2023
d2_2023_filtered = d2_2023[~d2_2023['country'].isin(countries_with_zero_entries_d2)]

In [111]:
# Create filtered single version for d2_2020
d2_2020_filtered = d2_2020[~d2_2020['country'].isin(countries_with_zero_entries_d2)]

# 03. Append D2 Datasets

In [112]:
# Concatenate D2 datasets for future visualisations
d2_full = pd.concat([d2_2023,d2_2022,d2_2021,d2_2020,d2_2019,d2_2018,d2_2017,d2_2016,d2_2015])

In [113]:
d2_full.head()

year  rank      country            regional indicator  ladder score  \
0  2023     1      Finland                Western Europe         7.804   
1  2023     2      Denmark                Western Europe         7.586   
2  2023     3      Iceland                Western Europe         7.530   
3  2023     4       Israel  Middle East and North Africa         7.473   
4  2023     5  Netherlands                Western Europe         7.403   

   log GDP per capita  social support  healthy life expectancy  \
0               1.888           1.585                    0.535   
1               1.949           1.548                    0.537   
2               1.926           1.620                    0.559   
3               1.833           1.521                    0.577   
4               1.942           1.488                    0.545   

   freedom to make life choices  generosity  perceptions of corruption  \
0                         0.772       0.126                      0.535   
1                         0.734       0.208                      0.525   
2                         0.738       0.250                      0.187   
3                         0.569       0.124                      0.158   
4                         0.672       0.251                      0.394   

   dystopia + residual  whisker-high  whisker-low  
0                2.363         7.875        7.733  
1                2.084         7.667        7.506  
2                2.250         7.625        7.434  
3                2.691         7.535        7.411  
4                2.110         7.460        7.346

In [114]:
d2_full.tail()

year  rank  country            regional indicator  ladder score  \
153  2015   154   Rwanda            Sub-Saharan Africa         3.465   
154  2015   155    Benin            Sub-Saharan Africa         3.340   
155  2015   156    Syria  Middle East and North Africa         3.006   
156  2015   157  Burundi            Sub-Saharan Africa         2.905   
157  2015   158     Togo            Sub-Saharan Africa         2.839   

     log GDP per capita  social support  healthy life expectancy  \
153            0.222078        0.773701                 0.428643   
154            0.286655        0.353862                 0.319098   
155            0.663199        0.474890                 0.721932   
156            0.015303        0.415867                 0.223961   
157            0.208676        0.139945                 0.284430   

     freedom to make life choices  generosity  perceptions of corruption  \
153                      0.592007    0.226279                   0.551911   
154                      0.484496    0.182600                   0.080100   
155                      0.156841    0.471790                   0.189060   
156                      0.118496    0.197269                   0.100620   
157                      0.364528    0.166809                   0.107313   

     dystopia + residual  whisker-high  whisker-low  
153             0.670418           NaN          NaN  
154             1.633279           NaN          NaN  
155             0.328579           NaN          NaN  
156             1.833019           NaN          NaN  
157             1.567258           NaN          NaN

In [126]:
d2_region = d2_full.drop(columns = ['year','rank','ladder score','log GDP per capita','social support','healthy life expectancy','freedom to make life choices','generosity','perceptions of corruption','dystopia + residual','whisker-high','whisker-low'])

In [127]:
d2_region.head()

country            regional indicator
0      Finland                Western Europe
1      Denmark                Western Europe
2      Iceland                Western Europe
3       Israel  Middle East and North Africa
4  Netherlands                Western Europe

In [128]:
d2_region = d2_region.rename(columns={'country': 'country name'})

In [131]:
# merge regional indicator into d1 dataset
d1_filtered = d1_filtered.merge(d2_region, on = ['country name'], how = 'left', indicator = True) 

In [132]:
d1_filtered

country name  year  life Ladder  log GDP per capita  social support  \
0         Albania  2017        4.640               9.455           0.638   
1         Albania  2017        4.640               9.455           0.638   
2         Albania  2017        4.640               9.455           0.638   
3         Albania  2017        4.640               9.455           0.638   
4         Albania  2017        4.640               9.455           0.638   
...           ...   ...          ...                 ...             ...   
4741     Zimbabwe  2022        3.296               7.670           0.666   
4742     Zimbabwe  2022        3.296               7.670           0.666   
4743     Zimbabwe  2022        3.296               7.670           0.666   
4744     Zimbabwe  2022        3.296               7.670           0.666   
4745     Zimbabwe  2022        3.296               7.670           0.666   

      healthy life expectancy at birth  freedom to make life choices  \
0                               69.050                         0.750   
1                               69.050                         0.750   
2                               69.050                         0.750   
3                               69.050                         0.750   
4                               69.050                         0.750   
...                                ...                           ...   
4741                            54.525                         0.652   
4742                            54.525                         0.652   
4743                            54.525                         0.652   
4744                            54.525                         0.652   
4745                            54.525                         0.652   

      generosity  perceptions of corruption  positive affect  negative affect  \
0         -0.031                      0.876            0.547            0.334   
1         -0.031                      0.876            0.547            0.334   
2         -0.031                      0.876            0.547            0.334   
3         -0.031                      0.876            0.547            0.334   
4         -0.031                      0.876            0.547            0.334   
...          ...                        ...              ...              ...   
4741      -0.070                      0.753            0.641            0.191   
4742      -0.070                      0.753            0.641            0.191   
4743      -0.070                      0.753            0.641            0.191   
4744      -0.070                      0.753            0.641            0.191   
4745      -0.070                      0.753            0.641            0.191   

              regional indicator _merge  
0     Central and Eastern Europe   both  
1     Central and Eastern Europe   both  
2     Central and Eastern Europe   both  
3     Central and Eastern Europe   both  
4     Central and Eastern Europe   both  
...                          ...    ...  
4741          Sub-Saharan Africa   both  
4742          Sub-Saharan Africa   both  
4743          Sub-Saharan Africa   both  
4744          Sub-Saharan Africa   both  
4745          Sub-Saharan Africa   both  

[4746 rows x 13 columns]

# 04. Export D2 Datasets

In [133]:
# Export dataframes to csv
d2_full.to_csv(os.path.join(path, '02 Data','Sourced Data', 'd2_full.csv'))

In [134]:
# Export filtered dataframe d1
d1_filtered.to_csv(os.path.join(path, '02 Data','Prepared Data', 'd1_filtered.csv'))

In [135]:
# Export filtered dataframe d2_comp
d2_comp_filtered.to_csv(os.path.join(path, '02 Data','Prepared Data', 'd2_comp_filtered.csv'))

In [136]:
# Export filtered dataframe d2_2023
d2_2023_filtered.to_csv(os.path.join(path, '02 Data','Prepared Data', 'd2_2023_filtered.csv'))

In [137]:
# Export filtered dataframe d2_2020
d2_2020_filtered.to_csv(os.path.join(path, '02 Data','Prepared Data', 'd2_2020_filtered.csv'))